java-checks/src/main/java/org/sonar/java/checks/TryWithResourcesCheck.java

```diff
@@ -55,10 +57,64 @@ public void leaveFile(JavaFileScannerContext context) {
   public void visitNode(Tree tree) {
     if (tree.is(Tree.Kind.TRY_STATEMENT)) {
       withinTry.push((TryStatementTree) tree);
-      toReport.push(new ArrayList<>());
-    } else if (withinStandardTryWithFinally() && ((NewClassTree) tree).symbolType().isSubtypeOf("java.lang.AutoCloseable")) {
-      toReport.peek().add(tree);
+      if (withinTry.size() != toReport.size()) {
+        toReport.push(new ArrayList<>());
+      }
+    } else if (((NewClassTree) tree).symbolType().isSubtypeOf("java.lang.AutoCloseable")) {
+      if (withinStandardTryWithFinally()) {
+        toReport.peek().add(tree);
+      } else if (isFollowedByTryWithFinally(tree)) {
+        if (toReport.isEmpty() || withinTry.size() == toReport.size()) {
+          // This newClass will be reported with the following tryStatement
+          toReport.push(new ArrayList<>());
+        }
+        toReport.peek().add(tree);
+      }
+    }
+  }
+
+  private static boolean isFollowedByTryWithFinally(Tree tree) {
+    Tree blockParent = tree.parent();
+    while (blockParent != null && !blockParent.is(Tree.Kind.BLOCK)) {
+      blockParent = blockParent.parent();
+    }
+
+    if (blockParent != null) {
+      CFG cfg = CFG.buildCFG(Collections.singletonList(blockParent));
+      if (!cfg.blocks().isEmpty()) {
+        return newFollowedByTryStatement(cfg.blocks().get(0));
+      }
+    }
+
+    return false;
```

Comment:
The 2 "return false" are unreachable by construction, so I would add a comment on top of each to make it explicit that it's defensive programming.